sled-storage/src/manager.rs

```diff
@@ -617,429 +836,520 @@ impl StorageManager {
+        let result = harness
+            .handle()
+            .omicron_physical_disks_ensure(config.clone())
+            .await
+            .expect("Ensuring disks should work after key manager is ready");
+        assert!(!result.has_error(), "{:?}", result);
 
-        // Spawn the key_manager so that it will respond to requests for encryption keys
-        tokio::spawn(async move { key_manager.run().await });
+        // Create a filesystem on the newly formatted U.2
+        let dataset_id = Uuid::new_v4();
+        let zpool_name = ZpoolName::new_external(config.disks[0].pool_id);
+        let dataset_name =
+            DatasetName::new(zpool_name.clone(), DatasetKind::Crucible);
+        harness
+            .handle()
+            .upsert_filesystem(dataset_id, dataset_name)
+            .await
+            .unwrap();
 
-        // Spawn the storage manager as done by sled-agent
-        tokio::spawn(async move {
-            manager.run().await;
-        });
+        harness.cleanup().await;
+        logctx.cleanup_successful();
+    }
 
-        handle.key_manager_ready().await;
+    #[tokio::test]
+    async fn ledgerless_to_ledgered_migration() {
+        illumos_utils::USE_MOCKS.store(false, Ordering::SeqCst);
+        let logctx = test_setup_log("ledgerless_to_ledgered_migration");
+        let mut harness = StorageManagerTestHarness::new(&logctx.log).await;
+
+        // Test setup: Create a U.2 and M.2
+        let raw_disks =
+            harness.add_vdevs(&["u2_under_test.vdev", "m2_helping.vdev"]).await;
+
+        // First, we format the U.2 to have a zpool. This should work, even
+        // without looping in the StorageManager.
+        let u2_raw = &raw_disks[0];
+        let pool_id = Uuid::new_v4();
+        let _disk = crate::disk::Disk::new(
+            &logctx.log,
+            &harness.mount_config(),
+            u2_raw.clone(),
+            Some(pool_id),
+            Some(harness.key_requester()),
+        )
+        .await
+        .expect("Failed to format U.2");
+
+        // Because we did that formatting "behind the back" of the
+        // StorageManager, we should see no evidence of the U.2 being managed.
+        //
+        // This currently matches the format of "existing systems, which were
+        // initialized before the storage ledger was created".
+
+        // We should still see no ledger.
+        let result = harness.handle().omicron_physical_disks_list().await;
+        assert!(matches!(result, Err(Error::LedgerNotFound),), "{:?}", result);
+
+        // We should also not see any managed U.2s.
+        let disks = harness.handle().get_latest_disks().await;
+        assert!(disks.all_u2_zpools().is_empty());
+
+        // However, when the system activates, we should see a Zpool,
+        // and "auto-manage" it.
+
+        harness.handle().key_manager_ready().await;
+
+        // It might take a moment for synchronization to be handled by the
```

Comment:
You can probably use `tokio::time::advance` here to speed things up. 10s is a long timeout.

Reply:
Good catch, updated